start_admin_sshd.sh

```diff
@@ -6,38 +6,94 @@ set -e
-  public_key_data="$(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}/${public_key_index}/openssh-key")"
-  if [[ ! "${public_key_data}" =~ ^"ssh" ]]; then
-    echo "Key ${public_key_data} with index ${public_key_index} looks invalid" >&2
-    continue
-  fi
-  echo "${public_key_data}" >> "${ssh_authorized_keys}"
-  if ! grep -q "${public_key_data}" "${ssh_authorized_keys}"; then
-    echo "Failed to write key with index ${public_key_index} to authorized_keys" >&2
-    continue
+declare -r ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
+declare -r ssh_config_dir="/home/ec2-user/.ssh"
+declare -r user_data="/.bottlerocket/host-containers/admin/user-data"
+available_auth_methods=0
+
+parse_authorized_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["authorized_keys"][]' "${user_data}" 2>/dev/null); then
+      echo "Failed to parse authorized_keys from ${user_data}" >&2
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        echo "Failed to validate ${key}" >&2
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate authorized_keys with all the authorized keys found in user-data
+if authorized_keys=$(parse_authorized_keys); then
+  ssh_authorized_keys="${ssh_config_dir}/authorized_keys"
+  install -D -m 0600 /dev/null "${ssh_authorized_keys}"
+  echo "${authorized_keys}" > "${ssh_authorized_keys}"
+  ((++available_auth_methods))
+fi
+
+parse_trusted_user_ca_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["trusted_user_ca_keys"][]' "${user_data}" 2>/dev/null); then
+      echo "Failed to parse trusted_user_ca_keys from ${user_data}" >&2
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        echo "Failed to validate ${key}" >&2
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate trusted_user_ca_keys with all the authorized keys found in user-data
+if trusted_user_ca_keys=$(parse_trusted_user_ca_keys); then
+  ssh_trusted_user_ca_keys="/etc/ssh/trusted_user_ca_keys.pub"
+  sshd_config="/etc/ssh/sshd_config"
+  install -D -m 0600 /dev/null "${ssh_trusted_user_ca_keys}"
+  echo "${trusted_user_ca_keys}" > "${ssh_trusted_user_ca_keys}"
+
+  cat <<EOF >> "${sshd_config}"
```

Comment:
Please put the heredoc starting token at the end of the redirections:

```suggestion
  cat >> "${sshd_config}" <<EOF
```